### DISCUSSION TITLE
General UX Feedback

### DISCUSSION THREAD

**Discussion by llibdude** [tinacms/tinacms#2630] [General]
I'm loving TinaCMS so far.  So much so that I made the time to create a feedback video using a test repo I whipped up.  The repo can be found here: https://github.com/llibdude/tinacmsplayground

I do apologize the video is 9 minutes but I will also annotate the video below.  Please split up my feedback as you see fit into features/bugs/ignore.

**I'm using a lot of experimental features, so keep that in mind as you watch.**

Here is the video on my gdrive. Quality isnt great i tried to make a smaller file size.
https://drive.google.com/file/d/1QmdVQ0ET5iXBl6Z-3fiDo5okIYdpuuse/view?usp=sharing

Annotations:
1:25 - Feature/Bug?: When I hit save on creating a new campaign I would expect to jump into the document, not the admin page again.  What is this supposed to do https://github.com/llibdude/tinacmsplayground/blob/main/pages/_app.tsx#L21? It seems like it should redirect me to my page but it does nothing (it doesnt even fire). I noticed it also doesnt work as i would expect for this repo https://github.com/tinacms/tinacms/tree/main/experimental-examples/tina-cloud-starter.

2:10 - Bug: Weird tab form focus issue
 
2:40 - Feature request: I would like default values for all the things, not just lists of objects or the `TinaTemplate` type.
Here is a small discord convo about this: https://discord.com/channels/835168149439643678/835172431836479508/948323585292922911

3:30 - Feature request: When adding a template, I would like to be navigated to the form of the template so that I can go ahead and edit that template with content.  Instead it kicks me back to the root of the document, and as a user I have to click more times to get back to the form.

~~4:00 - Feature request: `rich-text` type doesnt support a default value regardless. As I mention in the video I understand how this could be tricky to do.~~ nvm, see comment below

4:22 - Feature request: When adding things to a list, in my use case I want new things to be added to the end of the list not the front.  I understand there are other use cases where they need to be at the top.  A configuration option to control where new things in a list go would be great. Otherwise my users need to drag new templates to the bottom of the list after creating them.  Too many clicks.

6:00 - Feature request: Have some kind of naming function for template instances in a list so that I could change the name of each element in a list so that they can be unique from each other in a list for easier exploration.  I know the experimental hover/highlight feature exists which is certainly nice to have.

7:35 - Bug: Hitting tab on a field and hitting tab again on a template in a list that isnt the first breaks the UI.  You'll need to see the video it's hard to explain :).
After creating the video I perused the issues and found this related bug: https://github.com/tinacms/tinacms/issues/2626

8:20 - Assorted console errors that probably need clean up (I dont believe they are on my end)

Not in video:
- I would like to be able to have a block list that is not a list at all but just one item.  Already an issue filed here: https://github.com/tinacms/tinacms/issues/1862
- Cloudinary support.  The plugin could use a lot of improvements.
    - When I upload an image using the image field, it takes a long time and then when its complete the image doesnt save to the image field. I need to go into the media manager and reselect the image i just uploaded
    - Would be nice for organizational purposes if I had some control over where the image was saved to in cloudinary, instead of everything going into root. What could potentially be really nice is if images saved based on the collection or collection path. 
 
I am of course available to clarify any points.

**Comment by perkinsjr** [comment]
Thanks for the feedback! We will digest this and get back to you on your questions, feedback and ideas.

**Comment by scottgallant** [comment]
This is great!

Just a quick note about why we append items to the top of a list:  when appending to the bottom of the list, people don't see the list item get added when the list extends below the "fold" (below the viewport) and the user ends up clicking the add button multiple times because they don't think it works, not realizing it's appending to the bottom of the list (no visual feedback).  This often happens with long lists or lists that are close to the bottom of the page.

**Reply by llibdude** [reply]
Understood, given the choice for the particular use case in my video if there was a way to opt into opposite behavior I would use it.

**Reply by Enigmatical** [reply]
This is something for @spbyrne to consider, but I've seen this solved a couple of ways:

* An "Add to Bottom" button that sits alongside the default "Add" button (maybe smaller/minimal as it probably would not be the preferred "Add" button)
* "To Bottom" and "To Top" buttons in each List Item that lets you send them in a direction with a single click.

**Comment by llibdude** [comment]
I just realized I can set rich-text default values (in cases where default values work).  So you can disregard my feedback on that point.

```javascript
{
  ui: {
    defaultItem: {
       text: {
         children: [
          {
            type: 'p',
            children: [
              {
                type: 'text',
                text: 'Hero text content. This is a rich text component you can add hyperlinks, etc.'
              }
            ]
          }
        ]
      }
    }
  }
}
```

**Reply by jeffsee55** [reply]
👍 - We can probably allow you to set it as a raw markdown string to make it easier

**Reply by zeropaper** [reply]
In general, the default value concept is not easy to grasp, some more examples in the documentation (or better "highlighted") would be great :)

**Comment by Enigmatical** [comment]
> 1:25 - Feature/Bug?: When I hit save on creating a new campaign I would expect to jump into the document, not the admin page again. What is this supposed to do https://github.com/llibdude/tinacmsplayground/blob/main/pages/_app.tsx#L21? It seems like it should redirect me to my page but it does nothing (it doesn't even fire). I noticed it also doesn't work as I would expect for this repo https://github.com/tinacms/tinacms/tree/main/experimental-examples/tina-cloud-starter.

You've landed upon a problem that we currently do not have a good solution for (though, it has come up several times and I'm optimistic that we will).

Essentially, we cannot redirect you to the document you just created (in Production) because the document does not exist yet until the application has had a chance to rebuild.  `onNewDocument`, etc. is not supported by the CMS (it was a legacy feature of the old ContentCreator that had issues, as well).

For now, we navigate you back to the Document List rather than show a 404 - but neither is ideal.

**Reply by llibdude** [reply]
Perhaps when there are read tokens and nextjs's ISR can be used alongside TinaCMS this solves itself?

**Reply by Perni1984** [reply]
We've kind of "hacked" together a custom 404 page that "solves" this problem (until read tokens are available I guess).

The custom 404 page will use the useTina hook to query for the data for the newly created page (which only works if people are logged into tina cloud. If there is something returned other than `null` we are rendering the respective page with the respective data.

```typescript
import dynamic from 'next/dynamic';
import Error from 'next/error';
import { useRouter } from 'next/router';
import { useTina } from 'tinacms/dist/edit-state';
import { ErrorPage404 } from '../components/molecules/404ErrorPage/ErrorPage404';
import { CONSTANTS } from '../constants';
import { isBrowser } from '../shared/helper';
import {
  customGetPagesDocumentDocument,
  getBlogDocumentQuery,
} from '../shared/queries';

export const ClientSide404Error = () => {
  const queries = {
    '/': customGetPagesDocumentDocument,
    '/blog/': getBlogDocumentQuery,
  };

  const router = useRouter();
  const originalPathBefore404Error = router.asPath;

  const indexForSlicing = originalPathBefore404Error.lastIndexOf('/') + 1;
  const partialPath = originalPathBefore404Error.slice(0, indexForSlicing);

  let document = originalPathBefore404Error.slice(indexForSlicing);

  if (document === '') {
    document = CONSTANTS.indexPage;
  }
  const variables = {
    relativePath: `${document}.md`,
  };

  const { data } = useTina({
    query: queries?.[partialPath],
    variables,
    data: null,
  });

  if (!data) {
    return <ErrorPage404 />;
  }

  switch (partialPath) {
    case '/': {
      const ContentPage = dynamic(() =>
        import('../components/pages/ContentPage').then(
          (mod) => mod.ContentPage,
        ),
      );

      // @ts-ignore
      const props = data?.getPagesDocument?.data;

      return <ContentPage props={props} canonicalUrl={''} filename={''} />;
    }

    case '/blog/': {
      const BlogPage = dynamic(() =>
        import('../components/pages/BlogPage').then((mod) => mod.BlogPage),
      );

      // @ts-ignore - right now we don't have a way yet to generate types for custom queries
      const props = data?.getBlogDocument?.data;

      return <BlogPage props={props} canonicalUrl={''} cardReferences={[]} />;
    }

    default: {
      return <ErrorPage404 />;
    }
  }
};

export default function Page(props) {
  if (isBrowser()) {
    return <ClientSide404Error />;
  }
  return <ErrorPage404 />;
}

```

**Reply by llibdude** [reply]
read-only tokens were released this morning. I have confirmed that if you use them, this resolves the problem.

Now I understand that read-only tokens are experimental and opt-in, but since I am using them and they work, is there someway for me to opt into a direct navigation to a page after I create it?

**Comment by spbyrne** [comment]
@llibdude thanks for your feedback, it's exciting to see this type of use case! You make a lot of great points about UX.

- I agree default items would be great for groups, or in this case even defining a defaultItem at the page level could be useful.
- I think part of the reason we don't navigate you to a newly added block is that you don't have control of where it's added. I think if we add more control over where blocks are added it could be more intuitive to open the newly added form. Here's a thought around UI, how would you feel about something like this? 
    - <img width="802" alt="Screen Shot 2022-03-10 at 11 02 03 AM" src="https://user-images.githubusercontent.com/5075484/157689623-c97f8a9e-da7b-4b24-9dd1-07dd85e019fb.png">
- I agree having control over block and group labelling will significantly improve UX, especially for this type of application. The team's been working towards making the UI more extendible, and this is part of that work. I'm not sure on the timeline but I know it's coming.

**Reply by llibdude** [reply]
Hey @spbyrne thanks for the reply.  The image you shared to select where you want to add something in a list looks great to me. It would certainly solve my need and is likely the most graceful way to make all use cases happy.

> I think part of the reason we don't navigate you to a newly added block is that you don't have control of where it's added. I think if we add more control over where blocks are added it could be more intuitive to open the newly added form.

Thanks for giving me context as to why it behaves this way. I'm definitely biased here but in my use case seeing where in the list a block is added (which currently is always at the top anyway) is heavily outvalued by the pain I experience clicking an extra 20 times to create a simple page.
